assets/content/cookbook/Abstract types/abstracts-with-type-params.md

```diff
@@ -0,0 +1,158 @@
+function getProperty(property:String):Dynamic;
+
+// the expected `value` value depends on `property`
+function setProperty(property:String, value:Dynamic):Void;
+```
+
+For the Haxe code, this is considered a bad pattern, since it doesn't provide compile-time type checking and thus is unsafe.
+There are ways to do this safer, for example leveraging [ADT (enums)](http://haxe.org/manual/types-enum-instance.html),
+however we often need to provide extern definitions for some code that follows this pattern.
+
+Good news are - with Haxe we can stay type-safe by using type parameters and abstracts!
+
+Let's examine the `setProperty` example from the code above...
+
+
+## Step 1. Proper typing of string values
+
+Step 1 is to define a concrete type for our `property` argument, so we can't pass arbitrary string to it.
+But in run-time a string is still expected, so our type should be represented as strings in run-time, which is
+exactly what [abstract types](http://haxe.org/manual/types-abstract.html) are for:
+
+```haxe
+abstract Property(String) {
+  public inline function new(name) {
+    this = name;
+  }
+}
+```
+
+Then, let's change the `property` argument type to our new `Property` abstract type:
+
+```haxe
+function setProperty(property:Property, value:Dynamic):Void;
+```
+
+This already provides some compile-time safety because we need to have values of the `Property` type,
+not plain `String`, so we can't confuse it with other strings:
+
+```haxe
+var playerName = new Property("playerName"); // this could be some inlined constant
+
+setProperty(playerName, "Dan");
+
+// won't compile: String should be Property
+// setProperty("whatever", "Dan");
+```
+
+However we still can pass any value as the `value` argument, since it's typed `Dynamic`...
+
+
+## Step 2. Binding value type to a property name
+
+We want to associate our `Property` values with a concrete type for this property value.
+This is where [type parameters](http://haxe.org/manual/type-system-type-parameters.html) come into play.
+
+Let's parametrize our `Property` abstract type so its instances carry information about value type:
+
+```haxe
+abstract Property<T>(String) { // note the T type parameter
+  public inline function new(name) {
+    this = name;
+  }
+}
+```
+
+Now we can create "property" values specifying types of corresponding values, like this:
+
+```haxe
+// the value stored in playerName is String
+var playerName = new Property<String>("playerName");
+
+// playerLevel values are Int
+var playerLevel = new Property<Int>("playerLevel");
+
+// the info property is an object with given fields
+var info = new Property<{age:Int, dead:Bool}>("info");
+```
+
+Looks descriptive, but useless until we implement step 3...
+
+
+## Step 3. Using type parameter in a function
+
+Finally, let's make use of that type parameter `T` in our `setProperty` function.
+For that, we also parametrize the function and use the type parameter for the value instead of `Dynamic`:
```

Comment:
Well this doesn't look weird in my editor and it shouldn't make any difference for markdown, so whatever...